mPower2/mPower2/PassiveCollectors.swift

```diff
@@ -0,0 +1,291 @@
+        let timestamp = (location.timestamp as NSDate).iso8601String()!
+        let distance = location.distance(from: previousLocation)
+        let bearing = previousLocation.bearingInRadians(to: location)
+        let speed = (location.speed >= 0) ? location.speed : nil
+        let floor = location.floor?.level
+        let altitude = (location.verticalAccuracy >= 0) ? location.altitude : nil
+        let horizontalAccuracy = location.horizontalAccuracy
+        let verticalAccuracy = (location.verticalAccuracy >= 0) ? location.verticalAccuracy : nil
+        
+        // initialize dataPoint dict with the non-optional values
+        var dataPoint: [String: RSDJSONValue] = [
+            kTimestampKey: timestamp,
+            kDistanceKey: distance,
+            kBearingKey: bearing,
+            kHorizontalAccuracyKey: horizontalAccuracy
+        ]
+        
+        // now set the optional ones that are not nil
+        dataPoint[kSpeedKey] = speed
+        dataPoint[kFloorKey] = floor
+        dataPoint[kAltitudeKey] = altitude
+        dataPoint[kVerticalAccuracyKey] = verticalAccuracy
+        
+        return dataPoint
+    }
+    
+    func uploadDisplacement(_ displacementData: [String: RSDJSONValue]) {
+        let archiveFilename = kPassiveDisplacementArchiveFilename
+        let archive = SBBDataArchive(reference: schemaIdentifier, jsonValidationMapping: nil)
+        if let schemaRevision = MP2BridgeConfiguration.shared.schemaInfo(for: schemaIdentifier)?.schemaVersion {
+            archive.setArchiveInfoObject(schemaRevision, forKey: kSchemaRevisionKey)
+        }
+        archive.insertDictionary(intoArchive: displacementData, filename: archiveFilename, createdOn: Date())
+        
+        do {
+            try archive.complete()
+            archive.encryptAndUploadArchive()
+        }
+        catch {}
+   }
+    
+    // MARK: CLLocationManagerDelegate
+    func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
+        if status == .authorizedAlways {
+            // hey we have permission now, yay! so (re)start monitoring
+            start()
+        } else {
+            // we no longer have permission, but we can still upload what we collected while we did
+            stop(discarding: false)
+        }
+    }
+    
+    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
+        // Generate & upload (a) displacement(s) from the given location(s)
+        let validLocations = locations.filter({$0.horizontalAccuracy >= 0.0})
+        for location in validLocations {
+            if let lastLocation = previousLocation {
+                let dataPoint = displacementData(for: location, previousLocation: lastLocation)
+                uploadDisplacement(dataPoint)
+            }
+            previousLocation = location
```

Comment:
Is this thread safe? As in, will the location manager delegate always be called on the same thread? (I don't remember how it works).

Reply:
It's called on whatever thread you started location services on (in this case main).

I'll update the stop call to dispatch to main as well...